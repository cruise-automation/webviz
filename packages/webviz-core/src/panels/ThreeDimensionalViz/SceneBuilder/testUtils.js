// @flow
//
//  Copyright (c) 2021-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.
import type { Interactive } from "webviz-core/src/panels/ThreeDimensionalViz/Interactions/types";
import type { Marker } from "webviz-core/src/types/Messages";
import type { MarkerCollector } from "webviz-core/src/types/Scene";

export class OpenSourceMarkerCollector implements MarkerCollector {
  markers: Interactive<Marker>[] = [];
  grids: any[] = [];

  arrow = (o: any) => this.markers.push(o);
  cube = (o: any) => this.markers.push(o);
  cubeList = (o: any) => this.markers.push(o);
  sphere = (o: any) => this.markers.push(o);
  cylinder = (o: any) => this.markers.push(o);
  lineStrip = (o: any) => this.markers.push(o);
  lineList = (o: any) => this.markers.push(o);
  sphereList = (o: any) => this.markers.push(o);
  points = (o: any) => this.markers.push(o);
  text = (o: any) => this.markers.push(o);
  triangleList = (o: any) => this.markers.push(o);
  poseMarker = (o: any) => this.markers.push(o);
  filledPolygon = (o: any) => this.markers.push(o);
  instancedLineList = (o: any) => this.markers.push(o);
  overlayIcon = (o: any) => this.markers.push(o);
  grid(o: any) {
    this.grids.push(o);
  }
  pointcloud(o: any) {
    this.markers.push(o);
  }
  laserScan(_o: any) {
    throw new Error("not used in test");
  }
  radarPointCluster(_o: any) {
    throw new Error("not used in test");
  }
  linedConvexHull(_o: any) {
    throw new Error("not used in test");
  }
}
