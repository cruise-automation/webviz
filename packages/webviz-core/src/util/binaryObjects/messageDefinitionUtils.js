// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { sum } from "lodash";
import memoize from "memoize-weak";
import { type RosMsgField } from "rosbag";

import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";

const primitiveSizes = {
  json: 8,
  string: 8,
  bool: 1,
  int8: 1,
  uint8: 1,
  int16: 2,
  uint16: 2,
  int32: 4,
  uint32: 4,
  float32: 4,
  float64: 8,
  int64: 8,
  uint64: 8,
  time: 8,
  duration: 8,
};

export const primitiveList: Set<string> = new Set(Object.keys(primitiveSizes));

export class PointerExpression {
  variable: string;
  constant: number;

  constructor(variable: string, _constant: number = 0) {
    this.variable = variable;
    this.constant = _constant;
  }
  toString() {
    if (this.constant === 0) {
      return this.variable;
    }
    return `(${this.variable} + ${this.constant})`;
  }
  add(o: number) {
    return new PointerExpression(this.variable, this.constant + o);
  }
}

export const typeSize = memoize(
  (typesByName: RosDatatypes, typeName: string): number => {
    if (primitiveSizes[typeName] != null) {
      return primitiveSizes[typeName];
    }
    const messageType = typesByName[typeName];
    if (messageType == null) {
      throw new Error(`Unknown type: ${typeName}`);
    }
    // We add field sizes here, not type sizes. If the type has a field that's an array of strings,
    // it only adds eight bytes to the object's inline size.
    return sum(typesByName[typeName].fields.map((field) => fieldSize(typesByName, field)));
  }
);

// No point in memoizing this -- it can only do O(1) work before hitting `typeSize`, which is
// memoized. Memoizing on the string type name will work better than memoizing on field object
// identity, too.
export function fieldSize(typesByName: RosDatatypes, field: RosMsgField): number {
  if (field.isConstant) {
    return 0;
  }
  if (field.isArray) {
    // NOTE: Fixed-size arrays are not inlined.
    return 2 * primitiveSizes.int32;
  }
  return typeSize(typesByName, field.type);
}

// It's useful for field-accessor code for times and durations to share codepaths with field
// accessor code for complex message types. To do that, we add them to the set of messages.
// (Note, this might change depending on the code written next.)
export const addTimeTypes = (typesByName: RosDatatypes): RosDatatypes => ({
  ...typesByName,
  time: { fields: [{ name: "sec", type: "int32" }, { name: "nsec", type: "int32" }] },
  duration: { fields: [{ name: "sec", type: "int32" }, { name: "nsec", type: "int32" }] },
});

// String.prototype.replaceAll is not implemented in Chrome.
const allBadCharacters = new RegExp("[/-]", "g");
export const friendlyTypeName = (name: string): string => name.replace(allBadCharacters, "_");
export const deepParseSymbol = Symbol("deepParse");
type CommonBobjectSourceData = $ReadOnly<{|
  datatypes: RosDatatypes,
  datatype: string,
|}>;
export type BobjectSourceData =
  | CommonBobjectSourceData
  | $ReadOnly<{|
      ...CommonBobjectSourceData,
      buffer: ArrayBuffer, // Present in binary bobjects
      bigString: string,
      isArrayView: boolean,
    |}>;
export const classSourceData = new WeakMap<any, BobjectSourceData>();
export const associateSourceData = (cls: any, sourceData: BobjectSourceData): void => {
  classSourceData.set(cls, sourceData);
};
export const getSourceData = (cls: any): ?BobjectSourceData => classSourceData.get(cls);

// Note: returns false for "time" and "duration". Might be more useful not doing that.
// The RosMsgField flow-type has an `isComplex` field, but it isn't present for the websocket
// player, and has gone missing in test fixture before as well. Best to just assume it doesn't
// exist, because it's an annoying denormalization anyway.
export const isComplex = (typeName: string) => !primitiveList.has(typeName);

const numberList = ["int8", "uint8", "int16", "uint16", "int32", "uint32", "float32", "float64", "int64", "uint64"];
export const isNumberType = (typeName: string) => numberList.includes(typeName);
