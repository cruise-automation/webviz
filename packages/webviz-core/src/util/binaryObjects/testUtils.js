// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import type { RosValue } from "webviz-core/src/players/types";
import type { BinaryHeader } from "webviz-core/src/types/BinaryMessages";
import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";
import type { ArrayView } from "webviz-core/src/util/binaryObjects/ArrayViews";

export const definitions: RosDatatypes = {
  "std_msgs/Header": {
    fields: [{ type: "uint32", name: "seq" }, { type: "time", name: "stamp" }, { type: "string", name: "frame_id" }],
  },
  "fake_msgs/HasComplexAndArray": {
    fields: [{ type: "std_msgs/Header", name: "header" }, { type: "string", isArray: true, name: "stringArray" }],
  },
  "fake_msgs/HasComplexArray": {
    fields: [{ type: "fake_msgs/HasComplexAndArray", name: "complexArray", isArray: true }],
  },
  "fake_msgs/HasConstant": {
    fields: [{ type: "uint8", name: "const", isConstant: true, value: 1 }],
  },
  "fake_msgs/HasByteArray": {
    fields: [{ type: "uint8", name: "byte_array", isArray: true }],
  },
  "fake_msgs/HasJson": {
    fields: [{ type: "json", name: "jsonField" }],
  },
  "fake_msgs/HasInt64s": {
    fields: [{ type: "int64", name: "i64" }, { type: "uint64", name: "u64" }],
  },
  "fake_msgs/HasArrayOfEmpties": {
    fields: [{ type: "fake_msgs/HasConstant", name: "arr", isArray: true }],
  },
  "fake_msgs/HasBigIntArrays": {
    fields: [{ type: "int64", name: "i_arr", isArray: true }, { type: "uint64", name: "u_arr", isArray: true }],
  },
  "fake_msgs/ContainsEverything": {
    fields: [
      { type: "std_msgs/Header", name: "first" },
      { type: "fake_msgs/HasComplexAndArray", name: "second" },
      { type: "fake_msgs/HasComplexArray", name: "third" },
      { type: "fake_msgs/HasConstant", name: "fourth" },
      { type: "fake_msgs/HasByteArray", name: "fifth" },
      { type: "fake_msgs/HasJson", name: "sixth" },
      { type: "fake_msgs/HasInt64s", name: "seventh" },
      { type: "fake_msgs/HasArrayOfEmpties", name: "eighth" },
      { type: "fake_msgs/HasBigIntArrays", name: "ninth" },
    ],
  },
};

export type HasComplexAndArray = $ReadOnly<{|
  header(): BinaryHeader,
  stringArray(): ArrayView<string>,
|}>;

export type HasComplexArray = $ReadOnly<{|
  complexArray(): ArrayView<HasComplexAndArray>,
|}>;

export type HasConstant = $ReadOnly<{|
  const(): number,
|}>;

export type HasByteArray = $ReadOnly<{|
  byte_array(): Uint8Array,
|}>;

export type HasJson = $ReadOnly<{|
  jsonField(): RosValue,
|}>;

export type HasInt64s = $ReadOnly<{|
  i64(bigInt: ?true): number,
  u64(bigInt: ?true): number,
|}>;

export type HasArrayOfEmpties = $ReadOnly<{|
  arr(): ArrayView<HasConstant>,
|}>;

export type HasBigIntArrays = $ReadOnly<{|
  i_arr(): ArrayView<number>,
  u_arr(): ArrayView<number>,
|}>;

export type ContainsEverything = $ReadOnly<{|
  first(): BinaryHeader,
  second(): HasComplexAndArray,
  third(): HasComplexArray,
  fourth(): HasConstant,
  fifth(): HasByteArray,
  sixth(): HasJson,
  seventh(): HasInt64s,
  eighth(): HasArrayOfEmpties,
  ninth(): HasBigIntArrays,
|}>;
