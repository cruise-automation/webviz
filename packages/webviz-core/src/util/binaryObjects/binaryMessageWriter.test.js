// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { deepParse, getObject } from ".";
import {
  BobWriter,
  getSerializeFunctions,
  maybePrintStoreArray,
  printSerializationCode,
  printStoreMessageBody,
  printStoreSingularVariable,
} from "./binaryMessageWriter";
import { addTimeTypes, PointerExpression } from "./messageDefinitionUtils";
import { definitions } from "./testUtils";

describe("BobWriter", () => {
  it("can give an empty output", () => {
    const writer = new BobWriter();
    const { buffer, bigString } = writer.write();
    expect(bigString).toBe("");
    expect(buffer.byteLength).toBe(0);
  });

  it("can do simple writes", () => {
    const writer = new BobWriter();
    expect(writer.string("hello, ")).toBe(0);
    expect(writer.string("world")).toBe(7);
    let { offset, view } = writer.alloc(1);
    expect(offset).toBe(0);
    view.setUint8(offset, 1);
    ({ offset, view } = writer.alloc(1));
    expect(offset).toBe(1);
    view.setUint8(offset, 2);
    const { buffer, bigString } = writer.write();
    expect(bigString).toBe("hello, world");
    expect([...new Uint8Array(buffer)]).toEqual([1, 2]);
  });

  it("deduplicates strings", () => {
    const writer = new BobWriter();
    expect(writer.string("h")).toBe(0);
    expect(writer.string("ello")).toBe(1);
    expect(writer.string("ello")).toBe(1);
    const { buffer, bigString } = writer.write();
    expect(bigString).toBe("hello");
    expect(buffer.byteLength).toBe(0);
  });
});

describe("printStoreSingularVariable", () => {
  it("gives expected outputs", () => {
    const pointer = new PointerExpression("$offset").add(4);
    const variableName = "$field1";
    expect(printStoreSingularVariable(definitions, "json", variableName, pointer)).toBe(
      `const $field1$str = JSON.stringify($field1 === undefined ? null : $field1);
$view.setInt32(($offset + 4), $field1$str.length, true);
$view.setInt32(($offset + 8), $writer.string($field1$str), true);`
    );
    expect(printStoreSingularVariable(definitions, "string", variableName, pointer)).toBe(
      `const $field1$str = $field1 || "";
$view.setInt32(($offset + 4), $field1$str.length, true);
$view.setInt32(($offset + 8), $writer.string($field1$str), true);`
    );
    expect(printStoreSingularVariable(definitions, "bool", variableName, pointer)).toBe(
      "$view.setUint8(($offset + 4), +$field1);"
    );
    expect(printStoreSingularVariable(definitions, "int8", variableName, pointer)).toBe(
      "$view.setInt8(($offset + 4), $field1);"
    );
    expect(printStoreSingularVariable(definitions, "uint8", variableName, pointer)).toBe(
      "$view.setUint8(($offset + 4), $field1);"
    );
    expect(printStoreSingularVariable(definitions, "int16", variableName, pointer)).toBe(
      "$view.setInt16(($offset + 4), $field1, true);"
    );
    expect(printStoreSingularVariable(definitions, "uint16", variableName, pointer)).toBe(
      "$view.setUint16(($offset + 4), $field1, true);"
    );
    expect(printStoreSingularVariable(definitions, "int32", variableName, pointer)).toBe(
      "$view.setInt32(($offset + 4), $field1, true);"
    );
    expect(printStoreSingularVariable(definitions, "uint32", variableName, pointer)).toBe(
      "$view.setUint32(($offset + 4), $field1, true);"
    );
    expect(printStoreSingularVariable(definitions, "float32", variableName, pointer)).toBe(
      "$view.setFloat32(($offset + 4), $field1, true);"
    );
    expect(printStoreSingularVariable(definitions, "float64", variableName, pointer)).toBe(
      "$view.setFloat64(($offset + 4), $field1, true);"
    );
    expect(printStoreSingularVariable(definitions, "int64", variableName, pointer)).toBe(
      "$view.setBigInt64(($offset + 4), BigInt($field1 || 0), true);"
    );
    expect(printStoreSingularVariable(definitions, "uint64", variableName, pointer)).toBe(
      "$view.setBigUint64(($offset + 4), BigInt($field1 || 0), true);"
    );
    expect(() => printStoreSingularVariable(definitions, "time", variableName, pointer)).toThrow("unknown type");
  });
});

describe("maybePrintStoreArray", () => {
  it("prints expected code", () => {
    expect(maybePrintStoreArray(definitions, "string", "stringField", new PointerExpression("$offset").add(4))).toBe(
      `if (stringField == null) {
  $view.setFloat64(($offset + 4), 0, true);
} else {
  const stringField$l = stringField.length;
  let stringField$o = $alloc(stringField$l * 8);
  $view.setInt32(($offset + 4), stringField$l, true);
  $view.setInt32(($offset + 8), stringField$o, true);
  for (let stringField$i = 0; stringField$i < stringField$l; ++stringField$i) {
    const stringField$e = stringField[stringField$i];
    const stringField$e$str = stringField$e || "";
    $view.setInt32(stringField$o, stringField$e$str.length, true);
    $view.setInt32((stringField$o + 4), $writer.string(stringField$e$str), true);
    stringField$o += 8;
  }
}`
    );
  });
});

describe("printStoreMessageBody", () => {
  it("can print messages containing primitives", () => {
    expect(printStoreMessageBody(definitions, "fake_msgs/HasInt64s", "x", new PointerExpression("offset"))).toBe(
      `const v$i64 = x.i64;
$view.setBigInt64(offset, BigInt(v$i64 || 0), true);
const v$u64 = x.u64;
$view.setBigUint64((offset + 8), BigInt(v$u64 || 0), true);`
    );
  });

  it("can print messages containing complex children", () => {
    expect(
      printStoreMessageBody(definitions, "fake_msgs/HasComplexAndArray", "x", new PointerExpression("offset"))
    ).toBe(
      `const v$header = x.header;
if (v$header == null) {
  $storage.fill(0, offset, (offset + 20));
} else {
  $write$std_msgs_Header(v$header, offset);
}
const v$stringArray = x.stringArray;
if (v$stringArray == null) {
  $view.setFloat64((offset + 20), 0, true);
} else {
  const v$stringArray$l = v$stringArray.length;
  let v$stringArray$o = $alloc(v$stringArray$l * 8);
  $view.setInt32((offset + 20), v$stringArray$l, true);
  $view.setInt32((offset + 24), v$stringArray$o, true);
  for (let v$stringArray$i = 0; v$stringArray$i < v$stringArray$l; ++v$stringArray$i) {
    const v$stringArray$e = v$stringArray[v$stringArray$i];
    const v$stringArray$e$str = v$stringArray$e || "";
    $view.setInt32(v$stringArray$o, v$stringArray$e$str.length, true);
    $view.setInt32((v$stringArray$o + 4), $writer.string(v$stringArray$e$str), true);
    v$stringArray$o += 8;
  }
}`
    );
  });
});

describe("printSerializationCode", () => {
  it("returns the expected code", () => {
    expect(printSerializationCode(addTimeTypes(definitions))).toMatchSnapshot();
  });
});

const roundTrip = (writer, object, type) => {
  const functions = getSerializeFunctions(definitions, writer);
  functions[type](object);
  const { buffer, bigString } = writer.write();
  return deepParse(getObject(definitions, type, buffer, bigString));
};

describe("getSerializeFunction", () => {
  const writer = new BobWriter();
  beforeEach(() => {
    writer.write();
  });

  // TODO(steel): More tests -- round-tripping, absent fields etc.
  it("compiles", () => {
    const functions = getSerializeFunctions(definitions, writer);
    expect(Object.keys(functions)).toEqual([
      "std_msgs/Header",
      "fake_msgs/HasComplexAndArray",
      "fake_msgs/HasComplexArray",
      "fake_msgs/HasConstant",
      "fake_msgs/HasByteArray",
      "fake_msgs/HasJson",
      "fake_msgs/HasInt64s",
      "fake_msgs/HasArrayOfEmpties",
      "fake_msgs/HasBigIntArrays",
      "fake_msgs/ContainsEverything",
      "time",
      "duration",
    ]);
  });

  it("round-trips simple messages", () => {
    const parsedHeader = { seq: 1, stamp: { sec: 2, nsec: 3 }, frame_id: "f" };
    const roundTrippedHeader = roundTrip(writer, parsedHeader, "std_msgs/Header");
    expect(roundTrippedHeader).toEqual(parsedHeader);
  });

  it("handles missing singular fields", () => {
    const headerFromBlank = roundTrip(writer, {}, "std_msgs/Header");
    expect(headerFromBlank).toEqual({ seq: 0, frame_id: "", stamp: { sec: 0, nsec: 0 } });
  });

  it("handles missing arrays", () => {
    const hasArrayFromBlank = roundTrip(writer, {}, "fake_msgs/HasComplexArray");
    expect(hasArrayFromBlank).toEqual({ complexArray: [] });
  });

  it("handles missing JSON", () => {
    const hasJsonFromBlank = roundTrip(writer, {}, "fake_msgs/HasJson");
    expect(hasJsonFromBlank).toHaveProperty("jsonField");
    expect(hasJsonFromBlank.jsonField).toBeNull();
  });

  it("handles string arrays", () => {
    const stringArray = ["foo", "bar"];
    const hasArray = roundTrip(writer, { stringArray }, "fake_msgs/HasComplexAndArray");
    expect(hasArray.stringArray).toEqual(stringArray);
  });

  it("handles message arrays", () => {
    const element = { header: { stamp: { sec: 1, nsec: 2 }, seq: 3, frame_id: "a" }, stringArray: ["b", "c"] };
    const hasArray = { complexArray: [element, element] };
    expect(roundTrip(writer, hasArray, "fake_msgs/HasComplexArray")).toEqual(hasArray);
  });
});
