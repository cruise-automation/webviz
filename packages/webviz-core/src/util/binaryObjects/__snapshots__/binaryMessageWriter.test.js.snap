// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printSerializationCode returns the expected code 1`] = `
"const $functions = {};

let $view, $storage;
function $alloc(size) {
  const { view, storage, offset } = $writer.alloc(size);
  $view = view;
  $storage = storage;
  return offset;
}
$alloc(0);

function $write$std_msgs_Header($v, $o) {
  const v$seq = $v.seq;
  $view.setUint32($o, v$seq, true);
  const v$stamp = $v.stamp;
  if (v$stamp == null) {
    $storage.fill(0, ($o + 4), ($o + 12));
  } else {
    $write$time(v$stamp, ($o + 4));
  }
  const v$frame_id = $v.frame_id;
  const v$frame_id$str = v$frame_id || \\"\\";
  $view.setInt32(($o + 12), v$frame_id$str.length, true);
  $view.setInt32(($o + 16), $writer.string(v$frame_id$str), true);
}
$functions[\\"std_msgs/Header\\"] = (message) => {
  const $ret = $alloc(20);
  $write$std_msgs_Header(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasComplexAndArray($v, $o) {
  const v$header = $v.header;
  if (v$header == null) {
    $storage.fill(0, $o, ($o + 20));
  } else {
    $write$std_msgs_Header(v$header, $o);
  }
  const v$stringArray = $v.stringArray;
  if (v$stringArray == null) {
    $view.setFloat64(($o + 20), 0, true);
  } else {
    const v$stringArray$l = v$stringArray.length;
    let v$stringArray$o = $alloc(v$stringArray$l * 8);
    $view.setInt32(($o + 20), v$stringArray$l, true);
    $view.setInt32(($o + 24), v$stringArray$o, true);
    for (let v$stringArray$i = 0; v$stringArray$i < v$stringArray$l; ++v$stringArray$i) {
      const v$stringArray$e = v$stringArray[v$stringArray$i];
      const v$stringArray$e$str = v$stringArray$e || \\"\\";
      $view.setInt32(v$stringArray$o, v$stringArray$e$str.length, true);
      $view.setInt32((v$stringArray$o + 4), $writer.string(v$stringArray$e$str), true);
      v$stringArray$o += 8;
    }
  }
}
$functions[\\"fake_msgs/HasComplexAndArray\\"] = (message) => {
  const $ret = $alloc(28);
  $write$fake_msgs_HasComplexAndArray(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasComplexArray($v, $o) {
  const v$complexArray = $v.complexArray;
  if (v$complexArray == null) {
    $view.setFloat64($o, 0, true);
  } else {
    const v$complexArray$l = v$complexArray.length;
    let v$complexArray$o = $alloc(v$complexArray$l * 28);
    $view.setInt32($o, v$complexArray$l, true);
    $view.setInt32(($o + 4), v$complexArray$o, true);
    for (let v$complexArray$i = 0; v$complexArray$i < v$complexArray$l; ++v$complexArray$i) {
      const v$complexArray$e = v$complexArray[v$complexArray$i];
      if (v$complexArray$e == null) {
        $storage.fill(0, v$complexArray$o, (v$complexArray$o + 28));
      } else {
        $write$fake_msgs_HasComplexAndArray(v$complexArray$e, v$complexArray$o);
      }
      v$complexArray$o += 28;
    }
  }
}
$functions[\\"fake_msgs/HasComplexArray\\"] = (message) => {
  const $ret = $alloc(8);
  $write$fake_msgs_HasComplexArray(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasConstant($v, $o) {

}
$functions[\\"fake_msgs/HasConstant\\"] = (message) => {
  const $ret = $alloc(0);
  $write$fake_msgs_HasConstant(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasByteArray($v, $o) {
  const v$byte_array = $v.byte_array;
  if (v$byte_array == null) {
    $view.setFloat64($o, 0, true);
  } else {
    const v$byte_array$l = v$byte_array.length;
    let v$byte_array$o = $alloc(v$byte_array$l * 1);
    $view.setInt32($o, v$byte_array$l, true);
    $view.setInt32(($o + 4), v$byte_array$o, true);
    for (let v$byte_array$i = 0; v$byte_array$i < v$byte_array$l; ++v$byte_array$i) {
      const v$byte_array$e = v$byte_array[v$byte_array$i];
      $view.setUint8(v$byte_array$o, v$byte_array$e);
      v$byte_array$o += 1;
    }
  }
}
$functions[\\"fake_msgs/HasByteArray\\"] = (message) => {
  const $ret = $alloc(8);
  $write$fake_msgs_HasByteArray(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasJson($v, $o) {
  const v$jsonField = $v.jsonField;
  const v$jsonField$str = JSON.stringify(v$jsonField === undefined ? null : v$jsonField);
  $view.setInt32($o, v$jsonField$str.length, true);
  $view.setInt32(($o + 4), $writer.string(v$jsonField$str), true);
}
$functions[\\"fake_msgs/HasJson\\"] = (message) => {
  const $ret = $alloc(8);
  $write$fake_msgs_HasJson(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasInt64s($v, $o) {
  const v$i64 = $v.i64;
  $view.setBigInt64($o, BigInt(v$i64 || 0), true);
  const v$u64 = $v.u64;
  $view.setBigUint64(($o + 8), BigInt(v$u64 || 0), true);
}
$functions[\\"fake_msgs/HasInt64s\\"] = (message) => {
  const $ret = $alloc(16);
  $write$fake_msgs_HasInt64s(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasArrayOfEmpties($v, $o) {
  const v$arr = $v.arr;
  if (v$arr == null) {
    $view.setFloat64($o, 0, true);
  } else {
    const v$arr$l = v$arr.length;
    let v$arr$o = $alloc(v$arr$l * 0);
    $view.setInt32($o, v$arr$l, true);
    $view.setInt32(($o + 4), v$arr$o, true);
    for (let v$arr$i = 0; v$arr$i < v$arr$l; ++v$arr$i) {
      const v$arr$e = v$arr[v$arr$i];
      if (v$arr$e == null) {
        $storage.fill(0, v$arr$o, v$arr$o);
      } else {
        $write$fake_msgs_HasConstant(v$arr$e, v$arr$o);
      }
      v$arr$o += 0;
    }
  }
}
$functions[\\"fake_msgs/HasArrayOfEmpties\\"] = (message) => {
  const $ret = $alloc(8);
  $write$fake_msgs_HasArrayOfEmpties(message, $ret);
  return $ret;
};
function $write$fake_msgs_HasBigIntArrays($v, $o) {
  const v$i_arr = $v.i_arr;
  if (v$i_arr == null) {
    $view.setFloat64($o, 0, true);
  } else {
    const v$i_arr$l = v$i_arr.length;
    let v$i_arr$o = $alloc(v$i_arr$l * 8);
    $view.setInt32($o, v$i_arr$l, true);
    $view.setInt32(($o + 4), v$i_arr$o, true);
    for (let v$i_arr$i = 0; v$i_arr$i < v$i_arr$l; ++v$i_arr$i) {
      const v$i_arr$e = v$i_arr[v$i_arr$i];
      $view.setBigInt64(v$i_arr$o, BigInt(v$i_arr$e || 0), true);
      v$i_arr$o += 8;
    }
  }
  const v$u_arr = $v.u_arr;
  if (v$u_arr == null) {
    $view.setFloat64(($o + 8), 0, true);
  } else {
    const v$u_arr$l = v$u_arr.length;
    let v$u_arr$o = $alloc(v$u_arr$l * 8);
    $view.setInt32(($o + 8), v$u_arr$l, true);
    $view.setInt32(($o + 12), v$u_arr$o, true);
    for (let v$u_arr$i = 0; v$u_arr$i < v$u_arr$l; ++v$u_arr$i) {
      const v$u_arr$e = v$u_arr[v$u_arr$i];
      $view.setBigUint64(v$u_arr$o, BigInt(v$u_arr$e || 0), true);
      v$u_arr$o += 8;
    }
  }
}
$functions[\\"fake_msgs/HasBigIntArrays\\"] = (message) => {
  const $ret = $alloc(16);
  $write$fake_msgs_HasBigIntArrays(message, $ret);
  return $ret;
};
function $write$fake_msgs_ContainsEverything($v, $o) {
  const v$first = $v.first;
  if (v$first == null) {
    $storage.fill(0, $o, ($o + 20));
  } else {
    $write$std_msgs_Header(v$first, $o);
  }
  const v$second = $v.second;
  if (v$second == null) {
    $storage.fill(0, ($o + 20), ($o + 48));
  } else {
    $write$fake_msgs_HasComplexAndArray(v$second, ($o + 20));
  }
  const v$third = $v.third;
  if (v$third == null) {
    $storage.fill(0, ($o + 48), ($o + 56));
  } else {
    $write$fake_msgs_HasComplexArray(v$third, ($o + 48));
  }
  const v$fourth = $v.fourth;
  if (v$fourth == null) {
    $storage.fill(0, ($o + 56), ($o + 56));
  } else {
    $write$fake_msgs_HasConstant(v$fourth, ($o + 56));
  }
  const v$fifth = $v.fifth;
  if (v$fifth == null) {
    $storage.fill(0, ($o + 56), ($o + 64));
  } else {
    $write$fake_msgs_HasByteArray(v$fifth, ($o + 56));
  }
  const v$sixth = $v.sixth;
  if (v$sixth == null) {
    $storage.fill(0, ($o + 64), ($o + 72));
  } else {
    $write$fake_msgs_HasJson(v$sixth, ($o + 64));
  }
  const v$seventh = $v.seventh;
  if (v$seventh == null) {
    $storage.fill(0, ($o + 72), ($o + 88));
  } else {
    $write$fake_msgs_HasInt64s(v$seventh, ($o + 72));
  }
  const v$eighth = $v.eighth;
  if (v$eighth == null) {
    $storage.fill(0, ($o + 88), ($o + 96));
  } else {
    $write$fake_msgs_HasArrayOfEmpties(v$eighth, ($o + 88));
  }
  const v$ninth = $v.ninth;
  if (v$ninth == null) {
    $storage.fill(0, ($o + 96), ($o + 112));
  } else {
    $write$fake_msgs_HasBigIntArrays(v$ninth, ($o + 96));
  }
}
$functions[\\"fake_msgs/ContainsEverything\\"] = (message) => {
  const $ret = $alloc(112);
  $write$fake_msgs_ContainsEverything(message, $ret);
  return $ret;
};
function $write$time($v, $o) {
  const v$sec = $v.sec;
  $view.setInt32($o, v$sec, true);
  const v$nsec = $v.nsec;
  $view.setInt32(($o + 4), v$nsec, true);
}
$functions[\\"time\\"] = (message) => {
  const $ret = $alloc(8);
  $write$time(message, $ret);
  return $ret;
};
function $write$duration($v, $o) {
  const v$sec = $v.sec;
  $view.setInt32($o, v$sec, true);
  const v$nsec = $v.nsec;
  $view.setInt32(($o + 4), v$nsec, true);
}
$functions[\\"duration\\"] = (message) => {
  const $ret = $alloc(8);
  $write$duration(message, $ret);
  return $ret;
};
return $functions;
"
`;
